pkg/policy/mapstate.go

```diff
@@ -292,23 +334,28 @@ func (keys MapState) deleteKeyWithChanges(key Key, owner MapStateOwner, adds, de
 // denyPreferredInsertWithChanges inserts a key and entry into the map by giving preference
 // to deny entries, and L3-only deny entries over L3-L4 allows.
 // Incremental changes performed are recorded in 'adds' and 'deletes', if not nil.
-func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStateEntry, adds, deletes Keys) {
+func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStateEntry, adds, deletes Keys, identities Identities) {
 	allCpy := allKey
 	allCpy.TrafficDirection = newKey.TrafficDirection
 	// If we have a deny "all" we don't accept any kind of map entry
 	if v, ok := keys[allCpy]; ok && v.IsDeny {
 		return
 	}
-
 	if newEntry.IsDeny {
-		// case for an existing allow L4-only and we are inserting deny L3-only
 		switch {
+		// case for inserting a deny L3-only
 		case newKey.DestPort == 0 && newKey.Nexthdr == 0 && newKey.Identity != 0:
 			l4OnlyAllows := MapState{}
+			// look for broad allows "all/unknown" identity and insert the same
+			// entry, but with the specifc L3-deny identity. This ensures that
+			// the broad allow is being overridden in the marginal circumstance of
+			// as subset of its allow being denied by another policy.
 			for k, v := range keys {
 				if newKey.TrafficDirection == k.TrafficDirection &&
 					!v.IsDeny &&
-					k.Identity == 0 && (k.DestPort != 0 || k.Nexthdr != 0) {
+					(k.Identity == 0 ||
+						netsContainAny(newEntry.getNets(identities, newKey.Identity), v.getNets(identities, k.Identity))) &&
```

Comment:
incidentally, you could probably fold the `k.Identity == 0` check into the `netContainsAny` because the numeric identity `0` should contain all identities.

Reply:
Moving that function to `pkg/ip` precludes that I think, but I see what you mean.